src/coreclr/System.Private.CoreLib/src/System/Reflection/Emit/ModuleBuilder.cs

```diff
@@ -222,6 +222,8 @@ private int GetTokenFromTypeSpec(byte[] signature, int length)
 
         private int GetTypeRefNested(Type type, Module? refedModule, string? strRefedModuleFileName)
         {
+            type = type.IsByRef ? type.GetElementType()! : type;
```

Comment:
What is the situation where this path gets exercised?

It does not look right to "forget" that the type is byref here. It means that the returned typeref is going to match the type that the caller asked for.

Reply:
This path is the root of all calls to resolve a typeref. From how I was reading this, the intent here is the actual typedef and not the typespec. Which means that the type's actual resolvable token is the important bit here not its level of indirection. Anytime the SRE tries to resolve a type to a typedef, this is the code path.

Reply:
The target of the call is https://github.com/dotnet/runtime/blob/66a4f235e9f82acdc5b94e49b47be6d873088fb8/src/coreclr/vm/commodule.cpp#L16-L28

Reply:
Upon referencing ECMA, the following line in II.22.39 is perhaps telling:

> Note that TypeSpec tokens can be used with any of the CIL instructions that take a TypeDef or TypeRef
token;

This would seem to imply the `GetTypeRef` function above should be updated. Thoughts?

Reply:
I think that the callers of GetTypeRefNested should be updated to deal with byref cases and emit type spec for those. The handling of byref cases should happen in the same spot as the handling of other cases that need type spec (pointers, arrays, ...).

For example, I think that the IsByRef check should be added here: https://github.com/dotnet/runtime/blob/e1b2c074a35e77dce393441b54d93110d0b4afcc/src/coreclr/System.Private.CoreLib/src/System/Reflection/Emit/ModuleBuilder.cs#L1091